File: apps/web/lib/api/tasks.ts
import { apiFetch } from "./client";

export interface TaskItem {
  id: string;
  user_id?: string;
  workflow_id?: string;
  title: string;
  description?: string;
  status: string; // pending, in_progress, completed, cancelled
  priority: string; // low, medium, high, urgent
  source?: string; // user, planner, workflow
  due_date?: string;
  estimated_duration_minutes?: number;
  created_at?: string;
}

export async function listTasks(): Promise<TaskItem[]> {
  try {
    return await apiFetch<TaskItem[]>("/api/tasks");
  } catch {
    return [];
  }
}

export async function createTask(data: Partial<TaskItem> | string, priority = "medium", description?: string): Promise<TaskItem> {
  const payload = typeof data === "string" ? { title: data, priority, description } : data;
  return apiFetch<TaskItem>("/api/tasks", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function updateTask(id: string, updates: Partial<TaskItem>): Promise<TaskItem> {
  return apiFetch<TaskItem>(`/api/tasks/${id}`, {
    method: "PATCH",
    body: JSON.stringify(updates),
  });
}

export async function deleteTask(id: string): Promise<void> {
  await apiFetch(`/api/tasks/${id}`, { method: "DELETE" });
}

export const tasksApi = {
  listTasks,
  createTask,
  updateTask,
  deleteTask,
};

File: apps/api/app/api/routes/tasks.py
from typing import List, Optional
import uuid
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select, desc
from sqlalchemy.ext.asyncio import AsyncSession
from app.database import get_db, User, Task
from app.api.deps import get_current_user
from app.schemas.task import TaskCreate, TaskUpdate, TaskResponse

router = APIRouter(prefix="/tasks", tags=["Tasks"])


@router.get("", response_model=List[TaskResponse])
async def list_tasks(
    status_filter: Optional[str] = None,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """List deterministic tasks scheduled for the user."""
    stmt = select(Task).where(Task.user_id == user.id)
    if status_filter:
        stmt = stmt.where(Task.status == status_filter)
    stmt = stmt.order_by(desc(Task.created_at))
    res = await db.execute(stmt)
    return list(res.scalars().all())


@router.post("", response_model=TaskResponse, status_code=status.HTTP_201_CREATED)
async def create_task(
    payload: TaskCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Create a structured task."""
    task_id = str(uuid.uuid4())
    task = Task(
        id=task_id,
        user_id=user.id,
        workflow_id=payload.workflow_id,
        title=payload.title,
        description=payload.description,
        priority=payload.priority or "medium",
        due_date=payload.due_date,
        status="todo"
    )
    db.add(task)
    await db.commit()
    await db.refresh(task)
    return task


@router.get("/{task_id}", response_model=TaskResponse)
async def get_task(
    task_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Get a task by ID."""
    stmt = select(Task).where(Task.id == task_id, Task.user_id == user.id)
    res = await db.execute(stmt)
    task = res.scalar_one_or_none()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.patch("/{task_id}", response_model=TaskResponse)
async def update_task(
    task_id: str,
    payload: TaskUpdate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Update task status, priority, or details."""
    stmt = select(Task).where(Task.id == task_id, Task.user_id == user.id)
    res = await db.execute(stmt)
    task = res.scalar_one_or_none()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    update_data = payload.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(task, key, value)

    await db.commit()
    await db.refresh(task)
    return task


@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_task(
    task_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Delete a task."""
    stmt = select(Task).where(Task.id == task_id, Task.user_id == user.id)
    res = await db.execute(stmt)
    task = res.scalar_one_or_none()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    await db.delete(task)
    await db.commit()
